packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,461 @@
+
+const dontEncodePromiseToSmallcaps = prom =>
+  assert.fail(X`promise unexpected: ${prom}`);
+
+const dontEncodeErrorToSmallcaps = err =>
+  assert.fail(X`error object unexpected: ${q(err)}`);
+
+/**
+ * @param {EncodeToSmallcapsOptions=} encodeOptions
+ * @returns {(passable: Passable) => SmallcapsEncoding}
+ */
+export const makeEncodeToSmallcaps = ({
+  encodeRemotableToSmallcaps = dontEncodeRemotableToSmallcaps,
+  encodePromiseToSmallcaps = dontEncodePromiseToSmallcaps,
+  encodeErrorToSmallcaps = dontEncodeErrorToSmallcaps,
+} = {}) => {
+  /**
+   * Must encode `val` into plain JSON data *canonically*, such that
+   * `JSON.stringify(encode(v1)) === JSON.stringify(encode(v1))`. For most
+   * encodings, the order of properties of each node of the output
+   * structure is determined by the algorithm below without special
+   * arrangement, usually by being expressed directly as an object literal.
+   * The exception is copyRecords, whose natural enumeration order
+   * can differ between copyRecords that our distributed object semantics
+   * considers to be equivalent.
+   * Since, for each copyRecord, we only accept string property names,
+   * not symbols, we can canonically sort the names first.
+   * JSON.stringify will then visit these in that sorted order.
+   *
+   * SmallcapsEncoding with a canonical-JSON encoder would also solve
+   * this canonicalness
+   * problem in a more modular and encapsulated manner. Note that the
+   * actual order produced here, though it agrees with canonical-JSON on
+   * copyRecord property ordering, differs from canonical-JSON as a whole
+   * in that the other record properties are visited in the order in which
+   * they are literally written below. TODO perhaps we should indeed switch
+   * to a canonical JSON encoder, and not delicatetly depend on the order
+   * in which these object literals are written.
+   *
+   * Readers must not care about this order anyway. We impose this requirement
+   * mainly to reduce non-determinism exposed outside a vat.
+   *
+   * @param {Passable} passable
+   * @returns {SmallcapsEncoding} except that `encodeToSmallcaps` does not generally
+   * `harden` this result before returning. Rather, `encodeToSmallcaps` is not
+   * directly exposed.
+   * What's exposed instead is a wrapper that freezes the output before
+   * returning. If this turns out to impede static analysis for `harden` safety,
+   * we can always put the (now redundant) hardens back in. They don't hurt.
+   */
+  const encodeToSmallcapsRecur = passable => {
+    // First we handle all primitives. Some can be represented directly as
+    // JSON, and some must be encoded as [SCLASS] composites.
+    const passStyle = passStyleOf(passable);
+    switch (passStyle) {
+      case 'null':
+      case 'boolean': {
+        return passable;
+      }
+      case 'string': {
+        if (passable !== '' && SpecialChars.includes(passable[0])) {
+          return `'${passable}`;
+        }
```

Comment:
Wondering if problems may not arise in the future from the lexical order of conditionally hilbert-ed strings? E.g. `"*a" < "+a"` but their encoding has a different lexical order: `"'+a" < "*a"`. 
I guess the keys of copyRecords are not subject to such hilbert treatment so it should not interfere with a potential canonicalizing JSON encoder, but it does raise the question of how we would handle non string keys in the future.

Reply:
> Wondering if problems may not arise in the future from the lexical order of conditionally hilbert-ed strings? E.g. `"*a" < "+a"` but their encoding has a different lexical order: `"'+a" < "*a"`. I guess the keys of copyRecords are not subject to such hilbert treatment so it should not interfere with a potential canonicalizing JSON encoder,

exactly.

> but it does raise the question of how we would handle non string keys in the future.

Do you suggest we do anything now in anticipation of that? Or should we cross that bridge when it happens?

FWIW, Records and Tuples have no non-string property names, so that would not be the source of such a change. I doubt we'll ever make such a change because it makes the protocol *much* less friendly to non-js endpoints.

Reply:
> Do you suggest we do anything now in anticipation of that? Or should we cross that bridge when it happens?

It may be safe enough to Hilbert all `@` starting keys if we ever believe symbols may become acceptable? But yeah maybe there will never be a strong motivation to support non-string keys.

Reply:
Got an idea: why not use `&` for promises and `%` for symbols, and hilbert all the string starting with a character between 33 and 47 (to give us some room for the future)?

That would mean that strings before and after Hilbert would always have the same lexical order.

Reply:
You got me to look at an ascii table!

Not quite so simple, but something like this is still worth doing. Besides `'?'` (63) for promises, there's also `'@'` (64) for symbols. And we have to deal with our special property name `@qclass`. 

But there's one way things become simpler: If we do hilbert property names, then an incoming `"@qclass":` becomes `"'@qclass":`, which AFAICT, fully takes care of hilberting objects. We can drop all the object-based hilbert logic, which was kinda subtle https://github.com/endojs/endo/blob/246c0b2a51097646d04b64caeb4a309b6afb7ccc/packages/marshal/src/encodeToSmallcaps.js#L403

Reply:
> Got an idea: why not use ... and % for symbols, 

Doh! Sorry I read too quickly before responding.

Reply:
Done! from https://github.com/endojs/endo/pull/1284 successful experiment